Add recipe for org-node

meedstrom opened the pull request:
### Brief summary of what the package does

How to summarize briefly... it's yet another zettelkasten/notetaking package, highly compatible with org-roam but the internals are very different.  

What excites me the most is its sheer speed on constrained systems.  I also like that it stays close to upstream Org things.

### Direct link to the package repository

https://github.com/meedstrom/org-node

### Your association with the package

I am the author.

### Relevant communications with the upstream package maintainer

**None needed**

### Checklist

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [] My elisp byte-compiles cleanly
- [x] I've used `M-x checkdoc` to check the package's documentation strings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)


There's a couple of warnings from the byte-compiler I can't fix.

meedstrom commented:
Removed the :version-regexp that opted out of Stable. I don't have tags anyway.

meedstrom commented:
Fixed a bunch of compile warnings.

riscy commented:
Thanks for this.  Quick first pass -

`org-node-backlink.el` with byte-compile using Emacs 29.4:
```
org-node-backlink.el:64:2: Warning: docstring has wrong usage of unescaped single quotes (use \= or different quoting)
```

`org-node-backlink.el` with checkdoc 0.6.2 (fix *within reason*):
```
org-node-backlink.el:0: The first line should be of the form: ";;; package --- Summary"
org-node-backlink.el:0: You should have a section marked ";;; Commentary:"
org-node-backlink.el:21: You should have a section marked ";;; Code:"
org-node-backlink.el:162: First line is not a complete sentence
org-node-backlink.el:318: First line is not a complete sentence
```

`org-node-obsolete.el` with checkdoc 0.6.2 (fix *within reason*):
```
org-node-obsolete.el:0: You should have a section marked ";;; Commentary:"
org-node-obsolete.el:6: You should have a section marked ";;; Code:"
org-node-obsolete.el:102: All variables and subroutines might as well have a documentation string
org-node-obsolete.el:108: All variables and subroutines might as well have a documentation string
org-node-obsolete.el:114: All variables and subroutines might as well have a documentation string
org-node-obsolete.el:120: All variables and subroutines might as well have a documentation string
```

`org-node-obsolete.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- org-node-obsolete.el#L100: Loading a package should rarely add hooks
```

`org-node-parser.el` with checkdoc 0.6.2 (fix *within reason*):
```
org-node-parser.el:0: You should have a section marked ";;; Commentary:"
org-node-parser.el:18: You should have a section marked ";;; Code:"
org-node-parser.el:268: First line is not a complete sentence
```

`org-node-parser.el` with package-lint 20240507.1722 and `package-lint-main-file` = "org-node.el":
```
4 issues found:
19:12: warning: This file is not in the `cl-lib' ELPA compatibility package: require `cl-lib' instead.
300:0: error: `org-node-parser--result:paths-types' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
301:0: error: `org-node-parser--result:found-links' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
302:0: error: `org-node-parser--result:problems' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
```

`org-node.el` with byte-compile using Emacs 29.4:
```
org-node.el:1723:26: Warning: `org-show-context' is an obsolete function (as of 9.6); use `org-fold-show-context' instead.
org-node.el:1724:26: Warning: `org-show-entry' is an obsolete function (as of 9.6); use `org-fold-show-entry' instead.
org-node.el:2270:6: Warning: (lambda (args) \...) quoted with ' rather than with #'
org-node.el:2277:6: Warning: (lambda (args) \...) quoted with ' rather than with #'
org-node.el:2284:6: Warning: (lambda (args) \...) quoted with ' rather than with #'
In org-node--add-series-to-dispatch:
org-node.el:2311:31: Warning: Unknown slot `:incompatible'
In org-node-rewrite-links-ask:
org-node.el:2782:20: Warning: `org-show-context' is an obsolete function (as of 9.6); use `org-fold-show-context' instead.
org-node.el:2783:20: Warning: `org-show-entry' is an obsolete function (as of 9.6); use `org-fold-show-entry' instead.
```

`org-node.el` with checkdoc 0.6.2 (fix *within reason*):
```
org-node.el:205: Lisp symbol `org-node-filter-fn' should appear in quotes
org-node.el:1746: Lisp symbol `org-node--id<>node' should appear in quotes
org-node.el:1841: Lisp symbol `org-node-capture-target' should appear in quotes
org-node.el:2901: Lisp symbol `org-id-locations-file' should appear in quotes
org-node.el:2789: `y-or-n-p' argument should end with "?"
```

`org-node.el` with package-lint 20240507.1722 and `package-lint-main-file` = "org-node.el":
```
3 issues found:
250:10: warning: Use variable `user-emacs-directory' or function `locate-user-emacs-file' instead of a literal path to the Emacs user directory or files.
1622:20: error: You should depend on (emacs "29.1") if you need `string-glyph-decompose'.
1625:20: error: You should depend on (emacs "29.1") if you need `string-glyph-compose'.
```

`org-node.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- org-node.el#L2045: Consider using %T instead of %H:%M:%S in time strings
```

Stderr output while compiling/loading:
```
org-node.el: Warning: (lambda (args) \...) quoted with ' rather than with #'
org-node.el: Warning: (lambda (args) \...) quoted with ' rather than with #'
org-node.el: Warning: (lambda (args) \...) quoted with ' rather than with #'
```

Package and license:
- Avoid specifying `:branch` except in unusual cases
- org-node-backlink.el: (no header)
- org-node-obsolete.el: "Outta sight I'm not tempted to clean them up"
  - See https://www.gnu.org/software/emacs/manual/html_node/elisp/Obsolete-Functions.html re: a formal way to handle this
- org-node-obsolete.el needs *formal* license boilerplate and/or an [SPDX-License-Identifier](https://spdx.dev/ids/)

meedstrom commented:
Oh, hey, [melpazoid](https://github.com/riscy/melpazoid), that'll be a good tool for next time, I can save you time ;)

I've fixed much of what you pasted. 

## Remaining niggles


```
org-node-backlink.el:64:2: Warning: docstring has wrong usage of unescaped single quotes (use \= or different quoting)
```

I saw this, but it's a mystery to me.  There is no docstring near [line 64](https://github.com/meedstrom/org-node/blob/8d9de6f3328dec9e3c759a971530f0a7e5727be3/org-node-backlink.el#L64), nor is there supposed to be.  Nor was I able to find an errant single-quote in any of the docstrings in the file.


```
300:0: error: `org-node-parser--result:paths-types' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
301:0: error: `org-node-parser--result:found-links' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
302:0: error: `org-node-parser--result:problems' contains a non-standard separator `:', use hyphens instead (see Elisp Coding Conventions).
```

In my subjective opinion, these names improve clarity where they are used.  It's not as if I am using a non-standard separator everywhere, which I believe is the situation that this advice was made for.  Is it OK to keep them?


```
org-node.el:1723:26: Warning: `org-show-context' is an obsolete function (as of 9.6); use `org-fold-show-context' instead.
org-node.el:1724:26: Warning: `org-show-entry' is an obsolete function (as of 9.6); use `org-fold-show-entry' instead.
```

I support Emacs 28 which does not come with Org 9.6, so I will use the old names.


```
org-node.el:2270:6: Warning: (lambda (args) \...) quoted with ' rather than with #'
org-node.el:2277:6: Warning: (lambda (args) \...) quoted with ' rather than with #'
org-node.el:2284:6: Warning: (lambda (args) \...) quoted with ' rather than with #'
```

This comes from lambdas in the [a Transient definition](https://github.com/meedstrom/org-node/blob/8d9de6f3328dec9e3c759a971530f0a7e5727be3/org-node.el#L2262-2301).  I believe I am following the Transient calling convention, whose manual [says we can use lambdas](https://magit.vc/manual/transient/Suffix-Specifications.html).  Actually, I am not getting those warnings, and I'll wager it's because I explicitly install Transient 0.7.4.  Emacs 29.4 ships with Transient 0.4.3.


```
org-node.el:205: Lisp symbol `org-node-filter-fn' should appear in quotes
org-node.el:1746: Lisp symbol `org-node--id<>node' should appear in quotes
org-node.el:1841: Lisp symbol `org-node-capture-target' should appear in quotes
org-node.el:2901: Lisp symbol `org-id-locations-file' should appear in quotes
org-node.el:2789: `y-or-n-p' argument should end with "?"
```

In principle yes, in these cases no.  I think it will be clear why, for whoever reads the source.

```
org-node.el:250:10: warning: Use variable `user-emacs-directory' or function `locate-user-emacs-file' instead of a literal path to the Emacs user directory or files.
```

Same as above. [See](https://github.com/meedstrom/org-node/blob/8d9de6f3328dec9e3c759a971530f0a7e5727be3/org-node.el#L250).

```
org-node.el:1622:20: error: You should depend on (emacs "29.1") if you need `string-glyph-decompose'.
org-node.el:1625:20: error: You should depend on (emacs "29.1") if you need `string-glyph-compose'.
```

There is already a guard clause that checks emacs-major-version.


## Package and license

> * Avoid specifying `:branch` except in unusual cases

OK, so I'll justify myself.  Let's see.  It's common to have a stable branch and a dev branch, right? (Or a stable branch and a bunch of feature branches)

It's up to the repo owner whether stable branch is default, or if the dev branch is default.  (Expressed another way, "main" means different things in different projects.)

Now, I'd like my **volatile experimental branch of unexploded ordnance** to be "main".  So please let me have `:branch` to avoid distributing broken things.

> * See https://www.gnu.org/software/emacs/manual/html_node/elisp/Obsolete-Functions.html re: a formal way to handle this

Perhaps the file "org-node-obsolete.el" was a bit misnamed.  That's what I started out doing, which is why it still had that name.

Thing is, those builtin obsoletions are just for communication with other devs, about obsolete functions and variables in a library.

E.g. package A depends on package B, and package B marks something as obsolete.  The developer of package A will see it and can update.

But if an user Zoe is using package A -- well, very few users compile their initfiles, so she never sees any warning about anything.

I'll rename this file to org-node-changes.el.  It warns the user about changes in *user options*.  It's actually an idea I'd like to see upstreamed into Emacs core.

riscy commented:
Thanks!  Here's the salient stuff -

> I saw this, but it's a mystery to me.

Sorry about this, I may have been checking an earlier version of the code.  The issue has moved to line 74 re: `define-globalized-minor-mode org-node-backlink-global-mode`.  This is because the macroexpansion is generating documentation:
```elisp
(progn
  (progn
    (put 'org-node-backlink-global-mode 'globalized-minor-mode t)
    :autoload-end (defvar-local org-node-backlink-mode-major-mode nil))
  (define-minor-mode org-node-backlink-global-mode
"Toggle Org-Node-Backlink mode in all buffers.
With prefix ARG, enable Org-Node-Backlink-Global mode if ARG is positive;
otherwise, disable it.

If called from Lisp, toggle the mode if ARG is `toggle'.
Enable the mode if ARG is nil, omitted, or is a positive number.
Disable the mode if ARG is a negative number.

Org-Node-Backlink mode is enabled in all buffers where `(lambda nil (when
(derived-mode-p 'org-mode) (org-node-backlink-mode)))' would do it.

See `org-node-backlink-mode' for more information on Org-Node-Backlink mode."
```
and it's the literal insertion of the lambda that's causing the byte-compile warning.  I would just replace the lambda with a named function, not only to solve the byte-compile warning but because it will lead to better auto-generated documentation anyway.

> Is it OK to keep [colons in function names]?

I hear you, but I'd prefer not; the fewer exceptions to the checklist tools (like package-lint) that I have to keep in my head as a reviewer, the easier it is for me.  But, if we can agree as a MELPA community that we want to start allowing/ignoring colons in names (and package-lint gets updated accordingly) then I'd have no problem with it (perhaps we should raise an issue?).  I won't *absolutely* insist though.

> maybe it's because I explicitly install Transient 0.7.4. Emacs 29.4 ships with Transient 0.4.3.

I confirmed I'm running with transient 0.7.4 but I can see how the macroexpansion is leading to quoted lambdas.  This isn't a big deal but you could solve it by using named functions instead.  Not a blocker.

> It's up to the repo owner whether stable branch is default, or if the dev branch is default. (Expressed another way, "main" means different things in different projects.)

The recipe file can accommodate anything, and if you insist then I won't block this.  But you are making an unusual and non-preferred choice: we prefer the default branch because the recipe is simpler, and you can rename (or change) your default branch without updating your recipe.  Moreover, you should also consider that people might want to install your code using tools like straight.el, and they will have to remember to include your "melpa" branch in their straight.el recipe (even though they may not be using MELPA at all). You are making things slightly more complicated for your downstream clients.

meedstrom commented:
>  Sorry about this, I may have been checking an earlier version of the code. The issue has moved to line 74

No worries, I've made a ton of commits since. That's why I had linked a pinned commit for your inspection :) Future-proof links!

Thanks for the hint about what's going on. Fixed!

Also fixed the colons in function names.

Also fixed the quoted lambdas.

So that leaves just one thing, the branch question...

> we prefer the default branch because the recipe is simpler, and you can rename (or change) your default branch without updating your recipe.

I do get Melpa's perspective.  If people have a "dist" branch and then delete it as a project restructuring, they may not realize that this breaks the Melpa recipe they submitted three years ago.

A good immunization is to literally call the branch "melpa". I would not delete a branch named that... unless smoking some seriously off-brand stuff.

> Moreover, you should also consider that people might want to install your code using tools like straight.el, and they will have to remember to include your "melpa" branch in their straight.el recipe (even though they may not be using MELPA at all). You are making things slightly more complicated for your downstream clients.

I think of it the other way around: it makes less work for whoever wants to make the occasional pull-request.

Users of Straight either type a recipe like this

    (use-package org-node)

in which case they get the Melpa recipe, or they specify the path to the repo directly like

    (use-package org-node :straight (org-node :type git :fetcher github :repo "meedstrom/org-node"))

but very few people will do that once they can just write the earlier form.  Those who do, are likely going to make a pull-request anyway, which should be done against the development branch, not the distribution one.

Another upside of separating "main" from "melpa" branches: it lets me update the README as seen on the Github landing page with less effort.

Anyway, I'm always up to discuss further, but my current stance is to insist on the branch I submitted.

riscy commented:
Okay, I admit to not being a frequent straight user and was imagining expressions like the second form you mentioned.  :)

Merging this now - thanks again!

meedstrom commented:
Awesome!!!
